wsmaster/che-core-api-factory-bitbucket-server/src/main/java/org/eclipse/che/api/factory/server/bitbucket/server/HttpBitbucketServerApi.java

```diff
@@ -0,0 +1,326 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.factory.server.bitbucket.server;
+
+import static java.time.Duration.ofSeconds;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.type.TypeFactory;
+import com.google.common.base.Charsets;
+import com.google.common.base.Strings;
+import com.google.common.io.CharStreams;
+import com.google.common.net.HttpHeaders;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.UncheckedIOException;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.ws.rs.core.MediaType;
+import org.eclipse.che.api.factory.server.scm.exception.ScmBadRequestException;
+import org.eclipse.che.api.factory.server.scm.exception.ScmCommunicationException;
+import org.eclipse.che.api.factory.server.scm.exception.ScmItemNotFoundException;
+import org.eclipse.che.api.factory.server.scm.exception.ScmUnauthorizedException;
+import org.eclipse.che.commons.subject.Subject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of @{@link BitbucketServerApi} that is using @{@link HttpClient} to communicate
+ * with Bitbucket Server.s
+ */
+public class HttpBitbucketServerApi implements BitbucketServerApi {
+
+  private static final ObjectMapper OM = new ObjectMapper();
+
+  private static final Logger LOG = LoggerFactory.getLogger(HttpBitbucketServerApi.class);
+  private final URI serverUri;
+  private final AuthorizationHeaderSupplier headerProvider;
+
+  public HttpBitbucketServerApi(
+      String serverUrl, AuthorizationHeaderSupplier authorizationHeaderSupplier) {
+    this.serverUri = URI.create(serverUrl);
+    this.headerProvider = authorizationHeaderSupplier;
+  }
+
+  @Override
+  public boolean isConnected(String bitbucketServerUrl) {
+    return serverUri.equals(URI.create(bitbucketServerUrl));
+  }
+
+  @Override
+  public BitbucketUser getUser(Subject cheUser)
+      throws ScmUnauthorizedException, ScmCommunicationException {
+    try {
+      Set<String> usersByName =
+          getUsers(cheUser.getUserName())
+              .stream()
+              .map(u -> u.getSlug())
+              .collect(Collectors.toSet());
+
+      Optional<BitbucketUser> currentUser = findCurrentUser(usersByName);
```

Comment:
10-90-X. I expect that the case when we will iterate over all users will be rare because in most of the cases we will iterate over the users who contain current che's username in profile.